sdk/pinocchio/src/sysvars/instructions.rs

```diff
@@ -0,0 +1,236 @@
+    /// currently executing `Transaction.
+    #[inline(always)]
+    pub fn get_instruction_relative(
+        &self,
+        index_relative_to_current: i64,
+    ) -> Result<IntrospectedInstruction, ProgramError> {
+        unsafe {
+            let current_index = self.load_current_index() as i64;
+            let index = current_index.saturating_add(index_relative_to_current);
+            if index < 0 {
+                return Err(SanitizeError::IndexOutOfBounds)
+                    .map_err(|_| ProgramError::InvalidArgument);
+            }
+
+            let num_instructions = u16::from_le(*(self.data.as_ptr() as *const u16));
+            if index >= num_instructions as i64 {
+                return Err(SanitizeError::IndexOutOfBounds)
+                    .map_err(|_| ProgramError::InvalidArgument);
+            }
+
+            Ok(self.deserialize_instruction(index as usize))
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a AccountInfo> for Instructions<Ref<'a, [u8]>> {
+    type Error = ProgramError;
+
+    fn try_from(account_info: &'a AccountInfo) -> Result<Self, Self::Error> {
+        if account_info.key() != &INSTRUCTIONS_ID {
+            return Err(ProgramError::UnsupportedSysvar);
+        }
+
+        Ok(Instructions {
+            data: account_info.try_borrow_data()?,
+        })
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, PartialEq, Eq)]
+pub struct IntrospectedInstruction<'a> {
+    pub raw: *const u8,
+    pub marker: PhantomData<&'a [u8]>,
+}
+
+impl<'a> IntrospectedInstruction<'a> {
+    /// Get the account meta at the specified index.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not verify if the index is out of bounds.
+    ///
+    /// It is typically used internally within the `get_account_meta_at` function, which
+    /// performs the necessary index verification. However, to optimize performance for users
+    /// who are sure that the index is in bounds, we have exposed it as an unsafe function.
+    pub unsafe fn get_account_meta_at_unchecked(&self, index: usize) -> &IntrospectedAccountMeta {
+        let offset = core::mem::size_of::<u16>() + (index * IntrospectedAccountMeta::LEN);
+        &*(self.raw.add(offset) as *const IntrospectedAccountMeta)
+    }
+
+    /// Get the account meta at the specified index.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`SanitizeError::IndexOutOfBounds`] if the index is out of bounds.
+    pub fn get_account_meta_at(
+        &self,
+        index: usize,
+    ) -> Result<&IntrospectedAccountMeta, SanitizeError> {
+        if index >= unsafe { u16::from_le(*(self.raw as *const u16)) } as usize {
+            return Err(SanitizeError::IndexOutOfBounds);
+        }
+
+        Ok(unsafe { self.get_account_meta_at_unchecked(index) })
+    }
+
+    /// Get the program ID of the `Instruction`.
+    pub fn get_program_id(&self) -> &Pubkey {
+        unsafe {
```

Comment:
Same here, we can't guarantee the alignment. Also, let's wrap only the unsafe code in "unsafe":
```rust
// SAFETY: The first 2 bytes represent the number of accounts in the instruction.
let num_accounts = u16::from_le_bytes(unsafe { *(self.raw as *const [u8; 2]) });

unsafe {
        &*(self.raw.add(
                size_of::<u16>() + num_accounts as usize * size_of::<IntrospectedAccountMeta>(),
        ) as *const Pubkey)
}
```